### ISSUE TITLE
JVM binding

### ISSUE THREAD

**Issue by keyboardDrummer** [tree-sitter/tree-sitter#207]
Hi there,

I'd like to use tree-sitter, but from a JVM language. Is there any chance of a JVM binding repo appearing any time soon?

Thanks,

Remy

**Comment by maxbrunsfeld** [comment]
It would definitely be a great addition to the project. I don't personally plan to create one in the short term, but would definitely be available to answer questions if you or anyone else wants to work on one. I think it should be totally doable, and it might be possible to take some ideas from the node.js binding.

**Comment by Chris2011** [comment]
That would be great, would like to use it inside java. So a java-tree-sitter binding here is needed, or Kotlin, Scala, whatever :D. Is there an CLI for using tree-sitter with default config smth like that?

**Comment by maxbrunsfeld** [comment]
There is a CLI, but I think `java-tree-sitter` would want to use the `tree-sitter` C library directly, via JNI.

**Comment by Chris2011** [comment]
Sounds legit to me.

**Comment by keyboardDrummer** [comment]
Let's be clear that this issue is about _JVM_ bindings for tree-sitter, so they would work for Java/Kotlin/Scala, not just for Java.

However, a major appeal of the JVM is platform portability, and losing this become of using JNI is quite a cost. This suggestion is quite a stretch, but have you considered writing this library in an even more portable language than C? Kotlin for example allows compiling to native executables, but also to C/C++, and to JVM bytecode. See [this page](https://kotlinlang.org/docs/reference/native-overview.html) for more information.

**Comment by Chris2011** [comment]
@keyboardDrummer I know, that's why I said: ...or kotlin, or scala or wahtever JVM. I only wanted to make my use-case clear.

**Comment by maxbrunsfeld** [comment]
Kotlin seems kinda cool, but no, I have no plans to rewrite Tree-sitter in kotlin. Out of curiosity, what platform are you thinking of where JNI doesn't work? JNI seems like a pretty standard thing to use; e.g. android relies on sqlite pretty heavily.

The other route that I could imagine going is compiling the Tree-sitter C code to WASM, since I believe there is already interest in running WASM on the JVM, e.g. https://github.com/cretz/asmble.

**Comment by keyboardDrummer** [comment]
> Out of curiosity, what platform are you thinking of where JNI doesn't work? JNI seems like a pretty standard thing to use; e.g. android relies on sqlite pretty heavily.

Well, correct me if I'm missing something, but packaging a .jar with tree-sitter compiled to one or multiple architectures would limit that .jar to executing only on those architectures. In that sense you lose the portability property of .jars which is that they can be executed on any machine that has a JVM.

I thought about the C->WASM->JVM route, but my feeling is that compiling a language with dynamic memory allocation to one with GC allocation creates fundamental performance problems. To take WASM to JVM as an example, if you call `grow_memory` in WASM, you have to allocate a new ByteBuffer in JVM and copy over all the old bytes. The author of Asemble himself notes that it should not be used in any production context.

In contrast, you can compile a GC language to a dynamic allocation language just fine, by adding the GC as part of the output. I'll admit though that this inflates the size of the executable.

I wonder whether a unicorn solution would be to have a language that does not allow pointer arithmetic, but still does manual memory management: something like Rust without the unsafe keyword. If there's no pointer arithmetic, it should be compile-able to GC languages, right? Rust code like [this](https://doc.rust-lang.org/rust-by-example/custom_types/structs.html) would easily compile to JVM bytecode, right? And obviously Rust also compiles to native/manual memory managed languages.

**Comment by stereobooster** [comment]
[Trying to compile it to WASM](https://github.com/stereobooster/tree-sitter-wasm), but I'm a newbie in this subject so I haven't got anywhere close so far

**Comment by maxbrunsfeld** [comment]
> Well, correct me if I'm missing something, but packaging a .jar with tree-sitter compiled to one or multiple architectures would limit that .jar to executing only on those architectures. In that sense you lose the portability property of .jars which is that they can be executed on any machine that has a JVM.

Yeah, you're right. It's a native library, so it does need to be compiled for each platform and architecture. This is a very common situation, but I can see how it might complicate a system that's otherwise distributed as pure JVM bytecode.

> In contrast, you can compile a GC language to a dynamic allocation language just fine, by adding the GC as part of the output.

This is true, but if you're embedding a library into an application whose runtime has its own GC (like Node.js, Haskell, Ruby, Python, Julia), it's highly undesirable to have to add a *second* GC into your application. I haven't really heard of anyone doing that in production code.

> To take WASM to JVM as an example, if you call grow_memory in WASM, you have to allocate a new ByteBuffer in JVM and copy over all the old bytes.

This doesn't seem like a big deal if you're using a normal amount of memory (e.g. ~20 syntax trees). It's an operation that you'll have to do only a few times, if ever, for the entire time that your app runs, and copying raw byte buffers is very fast - (gigabytes per second on modern hardware).

Still, it's good to know that asemble isn't intended for production use. It seems like JNI is still the most promising way to go about using a C library on the JVM.

**Comment by apmasell** [comment]
I've started to do a reimplementation in Java for a project I'm working on. I'm making it a bit more idiomatically Java (_e.g._, rather than having symbols as strings, they're a generic type). I was going to embed this in the software I'm working on, but I can create a separate library if interested.

**Comment by Chris2011** [comment]
A separate lib would be nice. Yes :)

**Comment by apmasell** [comment]
I've started on https://github.com/apmasell/treesitter4j
Porting help is appreciated.

**Comment by boggle** [comment]
Here's another suggestion, that's perhaps pratical: There's currently a tremendous amount of activity going on around compiling rust and mature, and perhaps rust is currently the most viable way to generate wasm. If the tree-sitter runtime would have been written in rust, making it accessible on the jvm via wasm might perhaps be a quite viable route. OTH a rust runtime would be accessible from C and other existing platforms, too.

**Comment by maxbrunsfeld** [comment]
Someone has worked on this now: https://github.com/JetBrains/jsitter. I'm not sure how mature it is, and I believe its API is a bit different from the Tree-sitter API, as they are working directly with Tree-sitter's internal data structures to create a more "functional" API. But it seems very cool. If you're interested, I'd encourage you to contribute to that project.

**Comment by Chris2011** [comment]
Nice, thx for the Info @maxbrunsfeld

**Comment by fcurts** [comment]
I just discovered https://github.com/serenadeai/java-tree-sitter
